igvc_navigation/src/navigation_client/navigation_client.cpp

```diff
@@ -0,0 +1,185 @@
+#include "navigation_client.h"
+#include <parameter_assertions/assertions.h>
+
+#include <robot_localization/navsat_conversions.h>
+#include <igvc_utils/NodeUtils.hpp>
+#include <igvc_utils/StringUtils.hpp>
+
+NavigationClient::NavigationClient()
+{
+  ros::NodeHandle private_nh("~");
+
+  assertions::getParam(private_nh, "reading_from_file", reading_from_file_);
+  assertions::getParam(private_nh, "/waypoint_file_path", waypoint_file_path_);
+  assertions::getParam(private_nh, "waypoint_radius", waypoint_radius_);
+
+  const double waiting_time = 5.0;
+  // wait for the action server to come up
+  while (!client.waitForServer(ros::Duration(waiting_time)))
+  {
+    ROS_INFO("Waiting for the move_base action server to come up");
+  }
+
+  rviz_sub_ = nh_.subscribe("/move_base_simple/goal", 1, &NavigationClient::rvizWaypointCallback, this);
+
+  // wait for the utm->odom transform to become available
+  while (!tf_listener_.waitForTransform("odom", "utm", ros::Time(0), ros::Duration(waiting_time)))
+  {
+    ROS_INFO_STREAM("utm->odom transform not found. waiting...");
+  }
+
+  ROS_INFO_STREAM("utm->odom transform found!");
+
+  if (reading_from_file_)
+  {
+    loadWaypointsFile();
+    for (const geometry_msgs::PointStamped& waypoint : waypoints_list_)
+    {
+      sendPointAsGoalAndWait(waypoint);
+    }
+  }
+  else
+  {
+    ROS_INFO_STREAM("Waiting for waypoints from rviz.");
+    ros::spin();
```

Comment:
I think it's a better practice to have the `ros::spin()` outside the constructor, as this means that you could potentially run multiple nodes if you wanted to.